packages/grid/x-data-grid/src/tests/filtering.DataGrid.test.tsx

```diff
@@ -1248,4 +1249,76 @@ describe('<DataGrid /> - Filter', () => {
       );
     }).not.to.throw();
   });
+
+  it('should update the filter model on columns change', () => {
+    const columns = [{ field: 'id' }, { field: 'brand' }];
+    const rows = [
+      { id: 0, brand: 'Nike' },
+      { id: 1, brand: 'Adidas' },
+      { id: 2, brand: 'Puma' },
+    ];
+    const onFilterModelChange = spy();
+
+    function Demo(props: Omit<DataGridProps, 'columns'>) {
+      return (
+        <div style={{ width: 300, height: 300 }}>
+          <DataGrid
+            autoHeight={isJSDOM}
+            columns={columns}
+            filterModel={{
+              items: [{ field: 'brand', operator: 'equals', value: 'Puma' }],
+            }}
+            onFilterModelChange={onFilterModelChange}
+            {...props}
+          />
+        </div>
+      );
+    }
+    const { setProps } = render(<Demo rows={rows} />);
+    expect(getColumnValues(1)).to.deep.equal(['Puma']);
+
+    setProps({ columns: [{ field: 'id' }] });
+    expect(getColumnValues(0)).to.deep.equal(['0', '1', '2']);
+    expect(onFilterModelChange.callCount).to.equal(1);
+    expect(onFilterModelChange.lastCall.firstArg).to.deep.equal({ items: [] });
+  });
+
+  // See https://github.com/mui/mui-x/issues/9204
+  it('should not clear the filter model when both columns and filterModel change', async () => {
+    const columns = [{ field: 'id' }, { field: 'brand' }];
+    const rows = [
+      { id: 0, brand: 'Nike' },
+      { id: 1, brand: 'Adidas' },
+      { id: 2, brand: 'Puma' },
+    ];
+
+    const onFilterModelChange = spy();
+
+    function Demo(props: Omit<DataGridProps, 'columns'>) {
+      return (
+        <div style={{ width: 300, height: 300 }}>
+          <DataGrid
+            autoHeight={isJSDOM}
+            columns={columns}
+            filterModel={{
+              items: [{ field: 'brand', operator: 'equals', value: 'Puma' }],
+            }}
+            onFilterModelChange={onFilterModelChange}
+            {...props}
+          />
+        </div>
+      );
+    }
+    const { setProps } = render(<Demo rows={rows} />);
+    expect(getColumnValues(1)).to.deep.equal(['Puma']);
+
+    setProps({
+      columns: [{ field: 'id' }],
+      filterModel: {
+        items: [{ field: 'id', operator: 'equals', value: '1' }],
```

Comment:
As the `filterModel` prop is updated, technically it's changed, shouldn't this emit a `filterModelChange` here? Or we only emit it once it's internally changed. (Same with `sortModel`)

Reply:
Good question!
In controlled mode, we don't call onChange handler if the new value is the same as the prop:
https://github.com/mui/mui-x/blob/a7b87a4be7c98dd5f28a0a51fce49e692af61c88/packages/grid/x-data-grid/src/hooks/core/useGridStateInitialization.ts#L61

https://github.com/mui/mui-x/blob/a7b87a4be7c98dd5f28a0a51fce49e692af61c88/packages/grid/x-data-grid/src/hooks/core/useGridStateInitialization.ts#L97-L103

Reply:
Just a guess, but could this by we are still experiencing an issue? Refresh once and the filters/sorts are missing. Refresh again and they are back.